nautobot/utilities/forms/fields.py

```diff
@@ -76,47 +77,67 @@ def __init__(self, from_form, *args, **kwargs):
             )
 
     def to_python(self, value):
-
-        records = []
+        if value is None:
+            return None
         reader = csv.reader(StringIO(value.strip()))
+        return parse_csv(reader)
 
-        # Consume the first line of CSV data as column headers. Create a dictionary mapping each header to an optional
-        # "to" field specifying how the related object is being referenced. For example, importing a Device might use a
-        # `site.slug` header, to indicate the related site is being referenced by its slug.
-        headers = {}
-        for header in next(reader):
-            if "." in header:
-                field, to_field = header.split(".", 1)
-                headers[field] = to_field
-            else:
-                headers[header] = None
+    def validate(self, value):
+        if value is None:
+            return None
+        headers, records = value
+        validate_csv(headers, self.fields, self.required_fields)
+
+        return value
 
-        # Parse CSV rows into a list of dictionaries mapped from the column headers.
-        for i, row in enumerate(reader, start=1):
-            if len(row) != len(headers):
-                raise forms.ValidationError(f"Row {i}: Expected {len(headers)} columns but found {len(row)}")
-            row = [col.strip() for col in row]
-            record = dict(zip(headers.keys(), row))
-            records.append(record)
+
+class CSVFileField(forms.FileField):
+    """
+    A FileField (rendered as a ClearableFileInput) which accepts a file containing CSV-formatted data. It returns
+    data as a two-tuple: The first item is a dictionary of column headers, mapping field names to the attribute
+    by which they match a related object (where applicable). The second item is a list of dictionaries, each
+    representing a discrete row of CSV data.
+
+    :param from_form: The form from which the field derives its validation rules.
+    """
+
+    def __init__(self, from_form, *args, **kwargs):
+
+        form = from_form()
+        self.model = form.Meta.model
+        self.fields = form.fields
+        self.required_fields = [name for name, field in form.fields.items() if field.required]
+
+        super().__init__(*args, **kwargs)
+
+        if not self.label:
+            self.label = "CSV File"
+        if not self.help_text:
+            self.help_text = (
+                "Select a CSV file to upload. It should contain column headers in the first row and use commas "
+                "to separate values. Multi-line data and values containing commas may be wrapped "
+                "in double quotes."
+            )
+
+    def to_python(self, file):
+        if file is None:
+            return None
+
+        file = super().to_python(file)
+        if not file.name.endswith(".csv"):
+            raise forms.ValidationError(f'File "{file.name}" is not a CSV file')
```

Comment:
Perhaps overly restrictive? I'm not sure it's necessary to mandate a `.csv` filename extension.